File: src/livox_pcd_projection/src/projectCloud.cpp
#include "livox_pcd_projection/projectCloud.hpp"

using namespace std;
using namespace cv;

float max_depth = 60;
float min_depth = 3;

cv::Mat src_img;
cv::Mat rectified_img;
cv_bridge::CvImagePtr cv_ptr(new cv_bridge::CvImage);
cv_bridge::CvImagePtr old_cv_ptr;

// function declaration
void imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg);

vector<livox_interfaces::msg::CustomMsg> lidar_datas; 
vector<projected_pt> projected_pts;
vision_msgs::msg::Detection2DArray detection_msg;
string intrinsic_path, extrinsic_path, camera_topic, lidar_topic, detection_topic;
int threshold_lidar, refresh_rate;  // number of pointcloud points projected onto image
bool debug; // switch for debug mode

using std::placeholders::_1;

LivoxProjectionNode::LivoxProjectionNode(): Node("Projection") {
    // declare parameters and their default value
    this->declare_parameter<std::string>("intrinsic_path", "");
    this->declare_parameter<std::string>("extrinsic_path", "");
    this->declare_parameter<std::string>("camera_topic", "");
    this->declare_parameter<std::string>("lidar_topic", "");
    this->declare_parameter<std::string>("detection_topic", "");
    this->declare_parameter<int>("lidar_threshold", 20000);
    this->declare_parameter<int>("refresh_rate", 10);
    this->declare_parameter<bool>("debug", false);

    // get parameters from launch file
    intrinsic_path = this->get_parameter("intrinsic_path").as_string();
    extrinsic_path = this->get_parameter("extrinsic_path").as_string();
    camera_topic = this->get_parameter("camera_topic").as_string();            
    lidar_topic = this->get_parameter("lidar_topic").as_string();
    detection_topic = this->get_parameter("detection_topic").as_string();
    threshold_lidar = this->get_parameter("lidar_threshold").as_int();
    refresh_rate = this->get_parameter("refresh_rate").as_int();
    debug = this->get_parameter("debug").as_bool();

    RCLCPP_INFO(this->get_logger(), "Extrinsic path: %s", intrinsic_path.c_str());
    RCLCPP_INFO(this->get_logger(), "Intrinsic path: %s", extrinsic_path.c_str());
    RCLCPP_INFO(this->get_logger(), "Camera topic: %s", camera_topic.c_str());
    RCLCPP_INFO(this->get_logger(), "Lidar topic: %s", lidar_topic.c_str());
    RCLCPP_INFO(this->get_logger(), "Detection topic: %s", detection_topic.c_str());
    RCLCPP_INFO(this->get_logger(), "Lidar threshold: %d", threshold_lidar);
    RCLCPP_INFO(this->get_logger(), "Refresh rate is set to: %d", refresh_rate);
    RCLCPP_INFO(this->get_logger(), "Debug mode: %s", debug ? "true":"false");

    // initialize publisher and subscriber
    chatter_pub = this->create_publisher<dist_msg::msg::Dist>("distances", 10);
    cloud_sub = this->create_subscription<livox_interfaces::msg::CustomMsg>(lidar_topic, 1, std::bind(&LivoxProjectionNode::cloudCallback, this, _1));
    detection_sub = this->create_subscription<vision_msgs::msg::Detection2DArray>(detection_topic, 1, std::bind(&LivoxProjectionNode::detectionsCallback, this, _1));
    
    message_filters::Subscriber<sensor_msgs::msg::Image> image_sub(this.get(), camera_topic, 1);
    message_filters::Subscriber<livox_interfaces::msg::CustomMsg> cloud_sub(this.get(), lidar_topic, 1);
    message_filters::Subscriber<vision_msgs::msg::Detection2DArray> detection_sub(this.get(), detection_topic, 1);
    typedef sync_policies::ApproximateTime<sensor_msg::msg::Image, livox_interfaces::msg::CustomMsg, vision_msgs::msg::Detection2DArray> MySyncPolicy;
    Synchronizer<MySyncPolicy> sync(MySyncPolicy(10), image_sub, cloud_sub, detection_sub);
    sync.registerCallback(std::bind(&LivoxProjectionNode::callback, _1, _2, _3));

}


void LivoxProjectionNode::callback(const sensor_msgs::msg::Image::ConstSharedPtr &image_msg, const livox_interfaces::msg::CustomMsg::ConstSharedPtr &cloud_msg, const vision_msgs::msg::Detection2DArray::ConstSharedPtr &detection_msg) {
    // do something
    RCLCPP_INFO(this->get_logger(), "callback");
}


// set the color by distance to the cloud
void LivoxProjectionNode::getColor(int &result_r, int &result_g, int &result_b, float cur_depth) {
    float scale = (max_depth - min_depth)/10;
    if (cur_depth < min_depth) {
        result_r = 0;
        result_g = 0;
        result_b = 0xff;
    }
    else if (cur_depth < min_depth + scale) {
        result_r = 0;
        result_g = int((cur_depth - min_depth) / scale * 255) & 0xff;
        result_b = 0xff;
    }
    else if (cur_depth < min_depth + scale*2) {
        result_r = 0;
        result_g = 0xff;
        result_b = (0xff - int((cur_depth - min_depth - scale) / scale * 255)) & 0xff;
    }
    else if (cur_depth < min_depth + scale*4) {
        result_r = int((cur_depth - min_depth - scale*2) / scale * 255) & 0xff;
        result_g = 0xff;
        result_b = 0;
    }
    else if (cur_depth < min_depth + scale*7) {
        result_r = 0xff;
        result_g = (0xff - int((cur_depth - min_depth - scale*4) / scale * 255)) & 0xff;
        result_b = 0;
    }
    else if (cur_depth < min_depth + scale*10) {
        result_r = 0xff;
        result_g = 0;
        result_b = int((cur_depth - min_depth - scale*7) / scale * 255) & 0xff;
    }
    else {
        result_r = 0xff;
        result_g = 0;
        result_b = 0xff;
    }
}

// copy the image ptr from subscribed image msg
void imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg){
    try {
        // ROS_INFO("Recieved image.");
        cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
    }
    catch (cv_bridge::Exception& e) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Could not convert from to 'bgr8'.");
    }
}

// read pointcloud from msg and save to 
void LivoxProjectionNode::cloudCallback(const livox_interfaces::msg::CustomMsg& msg) {
    // ROS_INFO("Pointcloud recieved.");
    lidar_datas.push_back(msg);
    // set decay rate
    if (lidar_datas.size() > threshold_lidar/24000 + 1) {
        // lidar_datas.clear();
        lidar_datas.erase(lidar_datas.begin());  // pop front
    }
}

// detection msg callback
void LivoxProjectionNode::detectionsCallback(const vision_msgs::msg::Detection2DArray& msg) {
    RCLCPP_INFO(this->get_logger(), "Recived detection message with stamp:  %i", msg.header.stamp.sec);
    detection_msg = msg;
}


int main(int argc, char** argv){
    rclcpp::init(argc, argv);
    auto p = std::make_shared<LivoxProjectionNode>();
    RCLCPP_INFO(p->get_logger(), "Initializing LivoxProjectionNode");
    rclcpp.spin(p);

    // vector<float> intrinsic;
    // getIntrinsic(p->intrinsic_path, intrinsic);
    // vector<float> distortion;
    // getDistortion(p->intrinsic_path, distortion);
    // vector<float> extrinsic;
    // getExtrinsic(p->extrinsic_path, extrinsic);

	// // set intrinsic parameters of the camera
    // RCLCPP_INFO(p->get_logger(), "Setting the instrinsic parameters of the camera");
    // cv::Mat cameraMatrix = cv::Mat::eye(3, 3, CV_64F);
    // cameraMatrix.at<double>(0, 0) = intrinsic[0];
    // cameraMatrix.at<double>(0, 2) = intrinsic[2];
    // cameraMatrix.at<double>(1, 1) = intrinsic[4];
    // cameraMatrix.at<double>(1, 2) = intrinsic[5];

	// // set radial distortion and tangential distortion
    // cv::Mat distCoeffs = cv::Mat::zeros(5, 1, CV_64F);
    // distCoeffs.at<double>(0, 0) = distortion[0];
    // distCoeffs.at<double>(1, 0) = distortion[1];
    // distCoeffs.at<double>(2, 0) = distortion[2];
    // distCoeffs.at<double>(3, 0) = distortion[3];
    // distCoeffs.at<double>(4, 0) = distortion[4];

    // // variable initialization
    // cv::Mat view, rview, map1, map2;
    // vector<float> distances; // stores distances of the lidar point cloud within the bounding box
    
    // // if (debug) {
    // //     rclcpp::Node::SharedPtr node = rclcpp::Node::make_shared("image_publisher");
    // //     image_transport::ImageTransport it(node);
    // //     image_transport::Publisher image_pub = it.advertise("projection_result", 1);
    // //     image_transport::Subscriber sub = it.subscribe(camera_topic, 1, imageCallback); // subscribe to camera node
    // // }

    // rclcpp::Rate loop_rate(refresh_rate);

    // int timeout_counter = 0;
    // int timeout = 5; // unit: seconds
    // dist_msg::msg::Dist distances_msg;
    
    // // if (debug) {
    // //     cv::Size imageSize; 
    // //     imageSize.width = 1448;
    // //     imageSize.height = 568;
    // //     cv::initUndistortRectifyMap(cameraMatrix, distCoeffs, cv::Mat(),cv::getOptimalNewCameraMatrix(cameraMatrix, distCoeffs, imageSize, 1, imageSize, 0), imageSize, CV_16SC2, map1, map2);
    // // }
    // // TODO: make logging when no data is coming nicer, change to try maybe

    // while (rclcpp::ok()) {
    //     // if (!cv_ptr->image.empty() && !lidar_datas.empty()) {
    //     if (!lidar_datas.empty()) {
    //         // cv::Size imageSize = cv_ptr->image.size();
    //         // RCLCPP_INFO_ONCE(p->get_logger(), "Rectifying camera distortion");
    //         // cv::initUndistortRectifyMap(cameraMatrix, distCoeffs, cv::Mat(),cv::getOptimalNewCameraMatrix(cameraMatrix, distCoeffs, imageSize, 1, imageSize, 0), imageSize, CV_16SC2, map1, map2);
    //         RCLCPP_INFO_ONCE(p->get_logger(), "Start pointcloud projection!");
    //         timeout_counter = 0; // reset timeout counter

    //         // src_img = cv_ptr->image;
    //         // if (old_cv_ptr != cv_ptr) {
            
    //             // cv::remap(src_img, src_img, map1, map2, cv::INTER_LINEAR);  // correct the distortion            
                
    //             // project the point cloud on to the image
    //         float x, y, z;
    //         float theoryUV[2] = {0, 0};
    //         for (unsigned int i = 0; i < lidar_datas.size(); ++i) {
    //             for (unsigned int j = 0; j < lidar_datas[i].point_num; ++j) {
    //                 x = lidar_datas[i].points[j].x;
    //                 y = lidar_datas[i].points[j].y;
    //                 z = lidar_datas[i].points[j].z;

    //                 getTheoreticalUV(theoryUV, intrinsic, extrinsic, x, y, z);
    //                 int u = floor(theoryUV[0] + 0.5);
    //                 int v = floor(theoryUV[1] + 0.5);

    //                 projected_pts.push_back({u, v, x}); // store projected points in list

    //                 // if (debug) {
    //                 //     int r, g, b;
    //                 //     p->getColor(r, g, b, x);
    //                 //     Point pt(u, v);
    //                 //     circle(src_img, pt, 1, Scalar(b, g, r), -1);
    //                 // }
    //             }
    //         }

    //         float avg; 
    //         // iterate over all the bounding boxes
    //         while (!detection_msg.detections.empty()) {
    //             // extract info from msg
    //             vision_msgs::msg::Detection2D detection = detection_msg.detections[0];
    //             vision_msgs::msg::ObjectHypothesisWithPose results = detection.results[0];
                
    //             // extract bounding box info
    //             vision_msgs::msg::BoundingBox2D bbox2d = detection.bbox;
    //             geometry_msgs::msg::Pose2D center = bbox2d.center;

    //             RCLCPP_INFO(p->get_logger(), "detection object class id: %s with score %f", results.hypothesis.class_id.c_str(), results.hypothesis.score);
    //             RCLCPP_INFO(p->get_logger(), "   position (%f, %f)", center.x, center.y);
    //             RCLCPP_INFO(p->get_logger(), "   size %f x %f", bbox2d.size_x, bbox2d.size_y);

    //             for (projected_pt pt: projected_pts) {
    //                 if ((center.x + bbox2d.size_x / 2) >= pt.u && (center.x - bbox2d.size_x / 2) <= pt.u && (center.y + bbox2d.size_y / 2) >= pt.v && (center.y - bbox2d.size_y / 2) <= pt.u) {
    //                     if (pt.dist != nan("")) {
    //                         distances.push_back(pt.dist);
    //                     }
    //                 }
    //             }

    //             avg = std::accumulate(distances.begin(), distances.end(), 0.0) / distances.size(); // calculate average distance for this boundingbox
    //             distances_msg.distances.push_back(avg);

    //             RCLCPP_INFO(p->get_logger(), "Average distance within in bounding box is %f", avg);
    //             // cv::Rect rect(center.x - bbox2d.size_x / 2, center.y - bbox2d.size_y / 2, bbox2d.size_x, bbox2d.size_y);
    //             // cv::rectangle(src_img, rect, cv::Scalar(0, 0, 255), 5); 
    //             distances.clear();
    //             detection_msg.detections.erase(detection_msg.detections.begin()); // pop front
    //         }

    //         // save distance data within the bounding box
    //         // if (rect_x + rect_width >= u && u >= rect_x && rect_y + rect_width >= v && v >= rect_y) {
    //         //     distances.push_back(x); // TODO: is x the distance to the object? 
    //         // }
            
    //         // draw bounding box
    //         // cv::Rect rect(rect_x, rect_y, rect_width, rect_height);
    //         // cv::rectangle(src_img, rect, cv::Scalar(0, 0, 255)); 

    //         // publish projected image
    //         // if (debug) {
    //             // cv_ptr->image = src_img;
    //             // image_pub.publish(cv_ptr->toImageMsg());
    //         // }
            
    //         // display average distance within the bounding box
    //         // float avg = std::accumulate(distances.begin(), distances.end(), 0.0) / distances.size();
    //         // RCLCPP_INFO(p->get_logger(), "Measured distance: %f", avg);

    //         p->chatter_pub->publish(distances_msg);
    //         // old_cv_ptr = cv_ptr;
    //         distances_msg.distances.clear();
    //     // }
    //     }
    //     else {
    //         RCLCPP_INFO_ONCE(p->get_logger(), "Waiting for lidar pointcloud data...");
    //         RCLCPP_INFO(p->get_logger(), "timeout counter: %d", timeout_counter);
    //         timeout_counter++;
    //         if (timeout_counter > timeout * refresh_rate) {
    //             RCLCPP_ERROR(p->get_logger(), "Missing image and/or lidar data!");
    //             return 1;
    //         }
    //     }
    //     // rclcpp::spin_some(node);
    //     rclcpp::spin_some(p);
    //     loop_rate.sleep();
    // }
}


// // get params from launch file

// // read image from msg and save to cv_ptr


// /* =========================== MAIN =========================== */
// int main(int argc, char **argv) {
//     rclcpp::init(argc, argv);
//     rclcpp::spin(std::make_shared<LivoxProjection>());
//     rclcpp::shutdown();
// }


// int main(){
//     cout << "Hello World!" << endl;
// }

// rclcpp::Node::SharedPtr node = rclcpp::Node::make_shared("image_publisher");

// image_transport::ImageTransport it(node);
// image_transport::Publisher image_pub = it.advertise("projection_result", 1);
// image_transport::Subscriber sub = it.subscribe("camera/image_0", 1, std::bind(&LivoxProjectionNode::imageCallback, this, _1)); // subscribe to camera node

File: src/livox_pcd_projection/livox_pcd_projection/projectCloud.py
import rclpy
import cv2 
from rclpy.node import Node
from vision_msgs.msg import (
    Detection2D,
    Detection2DArray,
    BoundingBox2D,
    ObjectHypothesisWithPose,
    ObjectHypothesis
)
from rclpy.parameter import Parameter
from cv_bridge import CvBridge
import message_filters
from sensor_msgs.msg import Image, CameraInfo
from livox_interfaces.msg import CustomMsg
from dist_msg.msg import Dist
import numpy as np
import cupy as cp
import time

class ProjectionNode(Node):
    def __init__(self):
        super().__init__('projection_node')
        self.declare_parameter('camera_topic', '/right_camera/image')
        self.declare_parameter('detection_topic', '/right_set/bbox')
        self.declare_parameter('lidar_topic', '/livox/stamped')
        self.declare_parameter('intrinsic_path', '/home/inspirationagx01/livox_projection/calibration_data/parameters/intrinsic.txt')
        self.declare_parameter('extrinsic_path', '/home/inspirationagx01/livox_projection/calibration_data/parameters/extrinsic.txt')
        self.declare_parameter('lidar_threshold', 20000)
        self.declare_parameter('refresh_rate', 30)
        self.declare_parameter('debug', False)

        self.camera_topic = self.get_parameter('camera_topic').get_parameter_value().string_value
        self.detection_topic = self.get_parameter('detection_topic').get_parameter_value().string_value
        self.lidar_topic = self.get_parameter('lidar_topic').get_parameter_value().string_value
        self.intrinsic_path = self.get_parameter('intrinsic_path').get_parameter_value().string_value
        self.extrinsic_path = self.get_parameter('extrinsic_path').get_parameter_value().string_value
        self.lidar_threshold = self.get_parameter('lidar_threshold').get_parameter_value().integer_value
        self.refresh_rate = self.get_parameter('refresh_rate').get_parameter_value().integer_value
        self.debug = self.get_parameter('debug').get_parameter_value().bool_value
        
        self.get_logger().info(f"Camera topic: {self.camera_topic}")
        self.get_logger().info(f"Detection topic: {self.detection_topic}")
        self.get_logger().info(f"Lidar topic: {self.lidar_topic}")
        self.get_logger().info(f"intrinsic file path: {self.intrinsic_path}")
        self.get_logger().info(f"extrinsic file path: {self.extrinsic_path}")
        self.get_logger().info(f"lidar threshold: {self.lidar_threshold}")
        self.get_logger().info(f"node refresh rate: {self.refresh_rate}")
        self.get_logger().info(f"debug mode: {self.debug}")
        self.get_logger().info("Finished parsing parameters")

        self.load_extrinsic(self.extrinsic_path)
        self.load_intrinsic_distortion(self.intrinsic_path)

        self.lidar_decay_list = []

        self.get_logger().info("Starting message filter subscribers and registering callbacks")
        
        # create publisher for distances
        self.publisher = self.create_publisher(Dist, '/distances', 10)

        # create subscriber to bbox and pointcloud
        self.detection_sub = message_filters.Subscriber(self, Detection2DArray, self.detection_topic)
        self.lidar_sub = message_filters.Subscriber(self, CustomMsg, self.lidar_topic)

        if self.debug:
            # also subscribe to image for visualization purposes
            self.camera_sub = message_filters.Subscriber(self, Image, self.camera_topic)
            self.ts = message_filters.ApproximateTimeSynchronizer([self.camera_sub, self.detection_sub, self.lidar_sub], 10, 0.1)
            self.ts.registerCallback(self.debug_callback)
        else :
            self.ts = message_filters.ApproximateTimeSynchronizer([self.detection_sub, self.lidar_sub], 10, 1)
            self.ts.registerCallback(self.callback)

    def debug_callback(self, camera_msg, detection_msg, lidar_msg):
        # TODO: create specific visualizations for debug mode
        self.callback(self, detection_msg, lidar_msg)
                
    def callback(self, detection_msg, lidar_msg):
        if len(self.lidar_decay_list) > self.lidar_threshold / 24000 + 1:
           self.lidar_decay_list.pop(0)
        self.lidar_decay_list.append(lidar_msg)

        # saving all the lidar points in to an array
        projected_points = np.array([])
        x, y, z = [], [], []
        for i in range(len(self.lidar_decay_list)):
            for j in range(self.lidar_decay_list[i].point_num):
                x.append(self.lidar_decay_list[i].points[j].x)
                y.append(self.lidar_decay_list[i].points[j].y)
                z.append(self.lidar_decay_list[i].points[j].z)

            x_temp, y_temp, z_temp = np.array(x)[:,None], np.array(y)[:,None], np.array(z)[:,None]
            ones = np.ones((x_temp.shape[0], 1))
            points = cp.asarray(np.hstack((x_temp, y_temp, z_temp, ones)))
            if projected_points.size == 0:
                projected_points = self.getTheoreticalUV(x, points)
            else:
                projected_points = np.hstack((projected_points, self.getTheoreticalUV(x, points)))

            x.clear()
            y.clear()
            z.clear()

        # calculate the average distance to the bounding boxes
        dist_msg = Dist()
        for detection in detection_msg.detections:
            # parse data from detection_msg
            dist_msg.obj_classes.append(detection.results[0].hypothesis.class_id)
            bbox = detection.bbox
            pose2d = bbox.center
            center_x, center_y = pose2d.x, pose2d.y
            width, height = bbox.size_x, bbox.size_y

            temp = []
            us = projected_points[1, :]
            vs = projected_points[2, :]
            shrink = 1 # edit this to shrink the bounding box for distance calculation
            right = center_x + width / 2 * shrink
            left = center_x - width / 2 * shrink
            top = center_y - height / 2 * shrink
            bottom = center_y + height / 2 * shrink
            temp = projected_points[0, (us >= left) & (us <= right) & (vs >= top) & (vs <= bottom)] # query points w/in bounding box

            temp = temp[~np.isnan(temp)]
            result = self.calc_dist(temp)
            dist_msg.distances.append(result)
        
        # publish Dist message
        dist_msg.header.stamp = self.get_clock().now().to_msg() # timestamp
        dist_msg.count = len(detection_msg.detections)
        self.publisher.publish(dist_msg)

    def calc_dist(dist_array):
        """Change this functiont to use different distance calculating methods"""
        return np.median(dist_array)
                                
    def load_extrinsic(self, extrinsic_path):
        file = open(extrinsic_path, 'r')
        lines = file.readlines()
        extrinsic = np.zeros((4, 4))
        for i, line in enumerate(lines):
            if i == 0:
                continue
            else:
                line = line.split('  ')
                extrinsic[i-1]= [float(x) for x in line]
    
        file.close()
        self.extrin = cp.asarray(extrinsic[:3, :])
        self.extrinsic = cp.asarray(extrinsic)
        
        self.get_logger().info('Loaded extrinsic data')

    def load_intrinsic_distortion(self, intrinsic_path):
        file = open(intrinsic_path, 'r')
        lines = file.readlines()
        intrinsic = np.zeros((3, 3))
        distortion = np.zeros((5, 1))
        skip = [0, 4, 5]
        for i, line in enumerate(lines):
            if i in skip:
                continue
            elif i < 4:
                line = line.split(' ')
                intrinsic[i-1]= [float(x) for x in line if x != '']
            else:
                line = line.split(' ')
                for j, dist in enumerate(line):
                    if dist != '':
                        distortion[j] = float(dist)
        self.intrinsic = cp.asarray(intrinsic)
        self.distortion = cp.asarray(distortion)
        file.close()
        self.get_logger().info('Loaded intrinsic data')

    def getTheoreticalUV(self, x, points):
        # intrinsic is 3x3
        # extrinsic is 3x4
        # m3 4x1
        result = cp.matmul(cp.matmul(self.intrinsic, self.extrin), points.T)
        
        depth = result[2,:]
        u = result[0,:] / depth
        v = result[1,:] / depth
        u, v = cp.asnumpy(u), cp.asnumpy(v)
        x = np.array(x)

        return np.vstack((x, u, v))


def main(args=None):
    rclpy.init(args=args)

    node = ProjectionNode()

    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()